src/libsyntax/parse/parser.rs

```diff
@@ -4311,12 +4313,13 @@ impl<'a> Parser<'a> {
             id: ast::DUMMY_NODE_ID,
             rules: s,
             span: lo.to(self.prev_span),
+            recovered,
         }))
     }
 
     /// Parse a statement, including the trailing semicolon.
     pub fn parse_full_stmt(&mut self, macro_legacy_warnings: bool) -> PResult<'a, Option<Stmt>> {
-        let mut stmt = match self.parse_stmt_(macro_legacy_warnings) {
+        let mut stmt = match self.parse_stmt_without_recovery(macro_legacy_warnings)? {
```

Comment:
This seems to be a change independent from adding the `recovered` flag to blocks?
From what I see in the test diffs I'd keep this recovery in place.

Reply:
The recovery happens in the callee already, and using `parse_stmt_` swallows errors that we do want to see in `parse_block_tail`.